crates/synthesis/src/discrete_basis/solovay_kitaev.rs

```diff
@@ -0,0 +1,383 @@
+    /// errors. If unsuccessful, this falls back onto constructing the standard U(2) matrix
+    /// representation and converting it to SO(3).
+    pub fn synthesize_std(
+        &self,
+        gate: &StandardGate,
+        params: &[Param],
+        recursion_degree: usize,
+    ) -> Result<CircuitData, DiscreteBasisError> {
+        let input_sequence = GateSequence::from_std(gate, params)?;
+        self.synthesize_sequence(&input_sequence, recursion_degree)
+    }
+
+    /// Run Solovay Kitaev given a [GateSequence].
+    ///
+    /// Mainly an internal helper for other runner methods.
+    fn synthesize_sequence(
+        &self,
+        sequence: &GateSequence,
+        recursion_degree: usize,
+    ) -> Result<CircuitData, DiscreteBasisError> {
+        // Run SK recursion.
+        let mut output = self.recurse(sequence, recursion_degree);
+
+        // Do minor optimizations on the output sequence, e.g. cancel gate-inverse-pairs, which
+        // can arise due to the recursions (even though they should not be present in the initial
+        // set of basic approximations).
+        output.inverse_cancellation();
+
+        // Compute the error in the SO(3) representation to return with the circuit.
+        let circuit = output.to_circuit(Some(sequence))?;
+        Ok(circuit)
+    }
+
+    /// Run a recursion step for a gate sequence, given a recursion degree.
+    ///
+    /// If the degree is 0 (recursion root), return the closest approximation in the set of
+    /// basic approximations. Otherwise, decompose the difference between the approximation
+    /// and the target unitary as balanced group commutator, and recurse on each element.
+    fn recurse(&self, sequence: &GateSequence, degree: usize) -> GateSequence {
+        // Recursion root: return the best approximation in the precomputed set.
+        if degree == 0 {
+            if self.do_checks {
+                math::assert_so3("Recursion root", &sequence.matrix_so3);
+            }
+            let basic_approximation = self
+                .basic_approximations
+                .query(sequence)
+                .expect("No basic approximation in root found")
+                .clone();
+            return basic_approximation;
+        }
+
+        // Find a basic approximation of the target sequence...
+        let u_n1 = self.recurse(sequence, degree - 1);
+
+        // ... and then improve the delta in between that approximation and the target.
+        let delta = sequence.matrix_so3 * u_n1.matrix_so3.transpose();
+        let (matrix_vn, matrix_wn) = group_commutator_decomposition(&delta, self.do_checks);
+        let vn = GateSequence::from_so3(&matrix_vn, self.do_checks);
+        let wn = GateSequence::from_so3(&matrix_wn, self.do_checks);
+
+        // Recurse on the group commutator elements.
+        let v_n1 = self.recurse(&vn, degree - 1);
+        let w_n1 = self.recurse(&wn, degree - 1);
```

Comment:
Here we would just call 
```rust
let v_n1 = self.recurse(&matrix_vn, degree - 1);
let w_n1 = self.recurse(&matrix_wn, degree - 1);
```
adn we wouldn't need to construct `vn` and `wn` in the first place